src/python/pants/engine/addressable.py

```diff
@@ -21,7 +21,7 @@ class Addresses(Collection[Address]):
 class AddressWithOrigin:
   """A BuildFileAddress along with the cmd-line spec it was generated from."""
   address: BuildFileAddress
-  origin: Spec
+  origin: Union[AddressSpec, FilesystemResolvedSpec]
```

Comment:
This is an important change. We do not want to allow `FilesystemIgnoreSpec` or `FilesystemGlobSpec` (which is unresolved).

Reply:
The mixing of inheritance/mixins and unions here feels vaguely odd... but I don't really have any other suggestions.